### PR TITLE
Deploy account RPC method

### PR DESCRIPTION
**Pull request by marioiordanov**
## Usage related changes

Users will be able to call starknet_addDeployAccountTransaction

## Checklist:

- [x] Applied formatting - `./scripts/format.sh`
- [x] No linter errors - `./scripts/clippy_check.sh`
- [x] Performed code self-review
- [x] Rebased to the last commit of the target branch (or merged it into my branch)
- [ ] Documented the changes
- [x] Linked the issues which this PR resolves
- [x] Updated the tests
- [x] All tests are passing - `cargo test`

### CODE REVIEW

File: crates/starknet/src/starknet/add_deploy_account_transaction.rs [hunk 1]
```diff
@@ -0,0 +1,175 @@
...
+            env!("CARGO_MANIFEST_DIR"),
+            "/test_artifacts/account_without_validations/account.json"
+        );
+        let contract_class = load_cairo_0_contract_class(account_json_path).unwrap();
+
+        let erc_20_contract = predeployed::create_erc20().unwrap();
+        erc_20_contract.deploy(&mut starknet.state).unwrap();
+
+        let acc = Account::new(
+            Felt::from(acc_balance.unwrap_or(100)),
```
Comment (FabijanC) [inline crates/starknet/src/starknet/add_deploy_account_transaction.rs:151]: I don't think we should default to 100, it's unclear from the tests (only clear once you read the `setup` function). Why is acc_balance even optional?

File: crates/starknet/src/starknet/add_deploy_account_transaction.rs [hunk 2]
```diff
@@ -0,0 +1,175 @@
...
+        (deploy_account_transaction.0.contract_address().clone()).try_into()?;
+
+    match deploy_account_transaction
+        .0
+        .execute(&mut starknet.state.pending_state, &starknet.block_context)
+    {
+        Ok(tx_info) => {
+            starknet.handle_successful_transaction(
+                &transaction_hash,
+                Transaction::DeployAccount(Box::new(deploy_account_transaction)),
```
Comment (FabijanC) [inline crates/starknet/src/starknet/add_deploy_account_transaction.rs:35]: Why do we need to `Box` it?

File: crates/starknet/src/starknet/add_deploy_account_transaction.rs [hunk 3]
```diff
@@ -0,0 +1,175 @@
...
+
+        let (txn_hash, _) = starknet.add_deploy_account_transaction(transaction).unwrap();
+        let txn = starknet.transactions.get_by_hash_mut(&txn_hash).unwrap();
+
+        assert_eq!(txn.status, TransactionStatus::Rejected);
+    }
+
+    #[test]
+    fn deploy_account_transaction_successful_execution() {
+        let (mut starknet, addr, fee_token_address) = setup(None);
```
Comment (FabijanC) [inline crates/starknet/src/starknet/add_deploy_account_transaction.rs:94]: Can you rename `addr` to reflect that it's actually an account address?

File: crates/starknet/src/starknet/add_deploy_account_transaction.rs [hunk 4]
```diff
@@ -0,0 +1,175 @@
...
+            Felt::from(acc_balance.unwrap_or(100)),
+            dummy_felt(),
+            dummy_felt(),
+            contract_class.generate_hash().unwrap(),
+            contract_class,
+            erc_20_contract.get_address(),
+        )
+        .unwrap();
+
+        acc.deploy(&mut starknet.state).unwrap();
```
Comment (FabijanC) [inline crates/starknet/src/starknet/add_deploy_account_transaction.rs:160]: Why are we deploying the account that we should later be deploying using the DeployAccount transaction?

File: crates/starknet/src/starknet/add_deploy_account_transaction.rs [hunk 5]
```diff
@@ -0,0 +1,175 @@
...
+    use crate::account::Account;
+    use crate::constants::{self, DEVNET_DEFAULT_CHAIN_ID};
+    use crate::starknet::{predeployed, Starknet};
+    use crate::traits::{Accounted, HashIdentifiedMut, StateChanger, StateExtractor};
+    use crate::transactions::deploy_account_transaction::DeployAccountTransaction;
+    use crate::utils::test_utils::dummy_felt;
+    use crate::utils::{get_storage_var_address, load_cairo_0_contract_class};
+
+    #[test]
+    fn deploy_account_transaction_should_fail_due_to_low_balance() {
```
Comment (FabijanC) [inline crates/starknet/src/starknet/add_deploy_account_transaction.rs:71]: I feel like these tests would have been simpler if we replaced them with e2e equivalents relying on BackgroundDevnet. Another benefit is that we would be testing the actual Devnet with actual requests. But that's just a thought, I believe the tests are valid as they are now, we don't yet have to make this replacement  (if we are even ever going to make it).

**Review by FabijanC — CHANGES_REQUESTED** [review]

File: crates/starknet/src/transactions/deploy_account_transaction.rs [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+        Ok(self.0.hash_value().clone().into())
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    #[test]
+    #[ignore]
+    fn correct_transaction_hash_computation() {
+        panic!("Transaction hash computation should be checked")
```
[the remark this answers is not in the document]
Comment (marioiordanov) [reply crates/starknet/src/transactions/deploy_account_transaction.rs:62]: My idea is to run periodically cargo test -- --ignored, then we will see what was not implemented

File: crates/starknet/src/starknet/add_deploy_account_transaction.rs (hunk 2 above)
Comment (marioiordanov) [reply crates/starknet/src/starknet/add_deploy_account_transaction.rs:35]: clippy made complaints about it

File: crates/starknet/src/starknet/add_deploy_account_transaction.rs (hunk 5 above)
Comment (FabijanC) [reply crates/starknet/src/starknet/add_deploy_account_transaction.rs:71]: You don't

File: crates/starknet/src/transactions/deploy_account_transaction.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (FabijanC) [reply crates/starknet/src/transactions/deploy_account_transaction.rs:62]: ok!

[the remark this answers is not in the document]
Comment (marioiordanov) [reply crates/starknet/src/transactions/deploy_account_transaction.rs]: Fixed in [33dd692](https://github.com/0xSpaceShard/starknet-devnet-rs/pull/73/commits/33dd692a30ff8355f693f5e65d40d155875c7fae)

File: crates/starknet/src/starknet/add_deploy_account_transaction.rs (hunk 4 above)
Comment (marioiordanov) [reply crates/starknet/src/starknet/add_deploy_account_transaction.rs:160]: It was easier to deploy the account and then use the class hash of already deployed account to construct the deploy transaction

Comment (marioiordanov) [reply crates/starknet/src/starknet/add_deploy_account_transaction.rs:160]: But when thinking of it, its actually easier to just declare the class hash

Comment (marioiordanov) [reply crates/starknet/src/starknet/add_deploy_account_transaction.rs]: Changed in [95a0224](https://github.com/0xSpaceShard/starknet-devnet-rs/pull/73/commits/95a022461e1aed1f0f5f3434844cbc9eb52b1fd9)

File: crates/starknet/src/starknet/add_deploy_account_transaction.rs (hunk 1 above)
Comment (marioiordanov) [reply crates/starknet/src/starknet/add_deploy_account_transaction.rs]: Removed in [95a0224](https://github.com/0xSpaceShard/starknet-devnet-rs/pull/73/commits/95a022461e1aed1f0f5f3434844cbc9eb52b1fd9)

File: crates/starknet-server/src/api/json_rpc/write_endpoints.rs [hunk 7]
```diff
@@ -34,6 +36,33 @@ impl JsonRpcHandler {
...
+            .await
+            .add_deploy_account_transaction(convert_to_deploy_account_transaction(
+                request,
+                chain_id.into(),
+            )?)
+            .map_err(|err| match err {
+                starknet_core::error::Error::StateError(
+                    starknet_in_rust::core::errors::state_errors::StateError::MissingClassHash(),
+                ) => ApiError::ClassHashNotFound,
+                _ => ApiError::RpcError(RpcError::invalid_request()),
```
[the remark this answers is not in the document]
Comment (marioiordanov) [reply crates/starknet-server/src/api/json_rpc/write_endpoints.rs]: resolved in [650a3cb](https://github.com/0xSpaceShard/starknet-devnet-rs/pull/73/commits/650a3cb70d1a1a520454d4997a871828b412c389)

File: crates/starknet/src/starknet/add_deploy_account_transaction.rs [hunk 8]
```diff
@@ -0,0 +1,161 @@
...
+
+        assert_eq!(txn.status, TransactionStatus::AcceptedOnL2);
+        assert_eq!(starknet.state.state.address_to_nonce.len(), accounts_before_deployment + 1);
+        let account_balance_after_deployment =
+            starknet.state.get_storage(balance_storage_key).unwrap();
+
+        assert!(account_balance_before_deployment > account_balance_after_deployment);
+    }
+
+    /// Initializes starknet with 1 account - account without validations
```
Comment (FabijanC) [inline crates/starknet/src/starknet/add_deploy_account_transaction.rs:135]: Also not in the comment that the erc20 is deployed and the gas price is set to 1

Comment (marioiordanov) [reply crates/starknet/src/starknet/add_deploy_account_transaction.rs]: [2428c54](https://github.com/0xSpaceShard/starknet-devnet-rs/pull/73/commits/2428c54771a90955d5119efbb4cf58e06449a7ae)

**Review by FabijanC — APPROVED** [review]

**Pull request merged**